File: javascript/observer/WeatherRama/observer/displays/StatisticsDisplay.js
import {Observer} from "../Observer.js";
import {WeatherStation} from "../../subject/WeatherStation.js";

export class StatisticsDisplay extends Observer {

    #maxTemperature = 0.0
    #minTemperature = 200
    #temperatureSum= 0.0
    #numReadings = 0

    #weatherStation = new WeatherStation()

    constructor(weatherStation) {
        super();
        this.#weatherStation = weatherStation
        this.#weatherStation.registerObserver(this)
    }

    update(temperature, humidity, pressure) {
        super.update(temperature, humidity, pressure);
        this.#temperatureSum += temperature
        this.#numReadings++

        if (temperature > this.#maxTemperature) { this.#maxTemperature = temperature }
        if (temperature < this.#minTemperature) { this.#minTemperature = temperature }

        this.display()
    }

    display() {
        console.log("StatisticsDisplay: \n" +
            "Avg temperature = " + (this.#temperatureSum/this.#numReadings) + "\n" +
            "Max temperature = " + this.#maxTemperature + "\n" +
            "Min temperature = " + this.#minTemperature + "\n"
        )
    }
}

File: ruby/observer/WeatherRama/observer/displays/statistics_display.rb
require './observer/display'
require './observer/observer'
require './subject/weather_station'

class StatisticsDisplay
  include Observer
  include Display

  def initialize(weather_station)
    @temperature_sum = 0
    @num_readings = 0
    @max_temperature = 0
    @min_temperature = 200
    @weather_station = WeatherStation.new
    @weather_station = weather_station
    @weather_station.register_observer(self)
  end

  def update(temperature, humidity, pressure)
    super
    @temperature_sum += temperature
    @num_readings += 1

    if temperature > @max_temperature
      @max_temperature = temperature
    end

    if temperature < @min_temperature
      @min_temperature = temperature
    end

    display

  end

  def display
    super
    avg = @temperature_sum / @num_readings
    puts("\nStatisticsDisplay:")
    puts("Avg temperature = " + avg.to_s)
    puts("Max temperature = " + @max_temperature.to_s)
    puts("Min temperature = " + @min_temperature.to_s)
  end

end